lib/writers.js

```diff
@@ -194,63 +202,56 @@ class WriteQueue extends events.EventEmitter {
 
   run() {
     if (!this.isRunning) {
-      this.process();
+      this.isRunning = true;
+      // Use nextTick to allow the queue to build up on the current phase
+      process.nextTick(() => this.process());
     }
   }
 
   process() {
-    const self = this;
-    utils.whilst(
-      function condition() {
-        return self.queue.length > 0;
-      },
-      function whileProcess(next) {
-        self.isRunning = true;
-        const buffers = [];
-        const callbacks = [];
-        let totalLength = 0;
-        while (totalLength < self.coalescingThreshold && self.queue.length > 0) {
-          const writeItem = self.queue.shift();
-          try {
-            const data = writeItem.operation.request.write(self.encoder, writeItem.operation.streamId);
-            totalLength += data.length;
-            buffers.push(data);
-            callbacks.push(writeItem.callback);
-          }
-          catch (err) {
-            writeItem.callback(err);
-            // Break and flush what we have
-            break;
-          }
-        }
-        if (buffers.length === 0) {
-          // No need to invoke socket.write()
-          return next();
-        }
-        // Before invoking socket.write(), mark that the request has been written to avoid race conditions.
-        for (let i = 0; i < callbacks.length; i++) {
-          callbacks[i]();
-        }
-        self.netClient.write(Buffer.concat(buffers, totalLength), function socketWriteCallback(err) {
-          if (err) {
-            self.setWriteError(err);
-          }
-          // Allow IO between writes
-          setImmediate(next);
-        });
-      },
-      function loopFinished() {
-        // The queue is now empty
-        self.isRunning = false;
+    this.netClient.cork();
```

Comment:
This seems like a good idea, especially since we bound what we are writing on `coalescingThreshold`